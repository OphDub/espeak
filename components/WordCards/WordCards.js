import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native';
import PropTypes from 'prop-types';
import Card from '../Card/Card';
import Progress from '../Progress/Progress';

export default class WordCards extends Component {
  constructor(props) {
    super(props);
    this.state = {
      stack: this.props.navigation.state.params,
      words: [],
      currentCardId: 0,
      isFinished: false,
    };
  }
  static navigationOptions = {
    title: 'translations'
  }

  componentDidMount = async () => {
    const { id } = this.state.stack;
    const jsonResponse = await this.getDeck(id);
    const words = jsonResponse.map((word, index) => {
      if (index === 0) {
        return {...word, isCurrent: true, isCompleted: false};
      }

      return {...word, isCurrent: false, isCompleted: false};
    });

    this.setState({ words });
  }

  getDeck = async (deckId) => {
    try {
      const promise = await fetch(`https://espeak-be.herokuapp.com/api/v1/words/${deckId}`);

      if (promise.status > 226) {
        throw new Error('We could not find your deck!');
      } else {
        return await promise.json();
      }
    } catch (error) {
      throw error;
    }
  }

  handleCardLoad = () => {
    if (this.state.words.length === 0) {
      return <Text>No Cards to Render</Text>;
    }

    const currentCard = this.state.words.find(card => card.isCurrent === true);
    const { handleHintPoints } = this.props.screenProps;

    if (!currentCard) {
      return (
        <View style={styles.endMsgCont}>
          <Text style={styles.endMsg}>Great job!</Text>
          <TouchableOpacity
            onPress={() => this.props.navigation.navigate('Decks')}
            style={styles.homeBtn}
          >
            <Text style={styles.homeBtnText}>BACK TO MY DECKS</Text>
          </TouchableOpacity>
        </View>
      );
    }

    return <Card
      word={currentCard}
      onCorrectAnswer={this.handleCorrectAnswer}
      onHintPoints={ handleHintPoints }
    />;
  }

  handleCorrectAnswer = (word) => {
    const wordIndex = this.state.words.indexOf(word);
    let isFinished = wordIndex === this.state.words.length - 1;
    const words = this.state.words.map((word, index) => {
      if (index === wordIndex) {
        return {...word, isCurrent: false};
      } else if (index === (wordIndex + 1)) {
        return {...word, isCurrent: true};
      }

      return word;
    });

    this.props.screenProps.handlePoints();

    if (isFinished) {
      this.props.screenProps.updateUser();
    }

    this.setState({
      words,
      currentCardId: wordIndex + 1
    });
  }

  render() {
    const { userPoints } = this.props.screenProps;

    return (
      <View style={styles.container}>
        <Progress
          stack={this.state.stack}
          userPoints={userPoints}
          currentCardId={this.state.currentCardId}
          wordTotal={this.state.words.length}
        />
        { this.handleCardLoad() }
      </View>
    );
  }
}

WordCards.propTypes = {
  screenProps: PropTypes.object,
  navigation: PropTypes.object
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'stretch',
  },
  endMsg: {
    color: '#272727',
    fontSize: 20,
    marginBottom: 20,
  },
  endMsgCont: {
    alignItems: 'center',
  },
  homeBtn: {
    backgroundColor: '#3AAFB9',
    padding: 10,
    borderRadius: 8,
    shadowColor: '#979797',
    shadowOpacity: 0.7,
    shadowOffset: { width: 2, height: 2 },
    shadowRadius: 4,
  },
  homeBtnText: {
    color: '#FFF',
  },
});
