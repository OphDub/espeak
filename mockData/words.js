const words = [
  {
    english: 'hi',
    spanish: 'hola',
    hint: 'oh-la',
    stack_id: 1
  },
  {
    english: 'bye',
    spanish: 'adios',
    hint: 'ah-dee-ohs',
    stack_id: 1
  },
  {
    english: 'good morning',
    spanish: 'buenos dias',
    hint: 'wenos dee-ahz',
    stack_id: 1
  },
  {
    english: 'see you later',
    spanish: 'hasta luego',
    hint: 'ah-stah loo-eh-go',
    stack_id: 1
  },
  {
    english: 'good afternoon',
    spanish: 'buenas tardes',
    hint: 'wenos tahr-days',
    stack_id: 1
  },
  {
    english: 'good night',
    spanish: 'buenas noches',
    hint: 'wenos no-chayz',
    stack_id: 2
  },
  {
    english: 'how are you?',
    spanish: 'como estas?',
    hint: 'coh-mo eh-stas',
    stack_id: 2
  },
  {
    english: 'I am fine',
    spanish: 'Estoy bien',
    hint: 'eh-stoy bee-in',
    stack_id: 2
  },
  {
    english: 'What is your name?',
    spanish: 'Como te llamas?',
    hint: 'coh-mo teh ya-mahs',
    stack_id: 2
  },
  {
    english: 'My name is Bob',
    spanish: 'Me llamo Bob',
    hint: 'meh, yamo bob',
    stack_id: 2
  },
  {
    english: 'apple',
    spanish: 'la manzana',
    hint: 'lah mahn-zah-na',
    stack_id: 3
  },
  {
    english: 'banana',
    spanish: 'el platano',
    hint: 'el plah-ta-no',
    stack_id: 3
  },
  {
    english: 'pepper',
    spanish: 'el pimiento',
    hint: 'el pee-mi-ehn-to',
    stack_id: 3
  },
  {
    english: 'orange',
    spanish: 'la naranja',
    hint: 'lah nar-ahn-ha',
    stack_id: 3
  },
  {
    english: 'cucumber',
    spanish: 'el pepino',
    hint: 'ehl peh-pee-no',
    stack_id: 3
  },
  {
    english: 'breakfast',
    spanish: 'el desayuno',
    hint: 'el des-ai-oo-no',
    stack_id: 4
  },
  {
    english: 'lunch',
    spanish: 'el almuerzo',
    hint: 'ehl all-mu-ehr-zo',
    stack_id: 4
  },
  {
    english: 'dinner',
    spanish: 'la cena',
    hint: 'lah ce-nah',
    stack_id: 4
  },
  {
    english: 'plate',
    spanish: 'el plato',
    hint: 'ehl plah-to',
    stack_id: 4
  },
  {
    english: 'food',
    spanish: 'la comida',
    hint: 'lah coh-mee-dah',
    stack_id: 4
  },
  {
    english: 'dance',
    spanish: 'bailar',
    hint: 'buy-lahr',
    stack_id: 5
  },
  {
    english: 'run',
    spanish: 'correr',
    hint: 'cohr-rer',
    stack_id: 5
  },
  {
    english: 'walk',
    spanish: 'caminar',
    hint: 'cah-mee-nahr',
    stack_id: 5
  },
  {
    english: 'jump',
    spanish: 'saltar',
    hint: 'sal-tar',
    stack_id: 5
  },
  {
    english: 'eat',
    spanish: 'comer',
    hint: 'coh-mehr',
    stack_id: 5
  }
] 

module.exports = words;
