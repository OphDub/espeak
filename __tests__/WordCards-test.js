import React from 'react';
import { shallow, configure } from 'enzyme';
import WordCards from '../components/WordCards/WordCards';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import words from '../mockData/words';
configure({ adapter: new Adapter() });

describe('WordCards', () => {
  let mockNavigation;
  let mockScreenProps;
  let mockDeck;
  let wrapper;

  beforeAll(() => {
    mockDeck = [ words[0], words[1], words[2], words[3], words[4] ];
  })
  
  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockDeck)
    }));
    mockNavigation = {
      state: {
        params: {
          id: 1,
          category: 'Basics 1',
        },
      },
    };

    mockScreenProps = {
      handlePoints: jest.fn(),
      updateUserPoints: jest.fn(),
      userPoints: 50,
    };

    wrapper = renderer.create(<WordCards navigation={mockNavigation} screenProps={mockScreenProps} />);

  }) 

  it('renders correctly', () => {
    const wrap = wrapper.toJSON(); 
    expect(wrap).toMatchSnapshot();
  })

  it('should call fetch on componentDidMount', () => {
    expect(global.fetch).toHaveBeenCalled(); 
  })

    
  it('should set state on componentDidMount', async () => {
    const instance = await wrapper.getInstance();
    const updatedMockDeck = mockDeck.map((word, index) => {
      if (index === 0) {
        return {...word, isCurrent: true, isCompleted: false};
      }

      return {...word, isCurrent: false, isCompleted: false};
    })

    await instance.componentDidMount();

    expect(instance.state.currentCardId).toEqual(0);
    expect(instance.state.stack).toEqual(mockNavigation.state.params);
    expect(instance.state.words).toEqual(updatedMockDeck);
  })

  describe('getDeck', () => {
    it('should return an array of cards if status code ok', () => {
      const instance = wrapper.getInstance();
      const response = instance.getDeck(1);
      const expected = mockDeck;

      expect(response).resolves.toEqual(expected);
    });

    it('should console error an error if fetch fails', () => {
      global.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 500
      }));
      const wrapper =  renderer.create(<WordCards navigation={mockNavigation} screenProps={mockScreenProps} />);
      const expected = Error('We could not find your deck!');
      const response = wrapper.getInstance().getDeck(1);

      expect(response).rejects.toEqual(expected);
    });
  })

  describe('handleCardLoad', () => {
    it('should return text saying "No Cards to Render" if no words are stored in state', () => {
      const instance = wrapper.getInstance();

      instance.setState({ words: [] });

      expect(wrapper.toJSON()).toMatchSnapshot();
    })

    it('should return end screen if there is no current card left in deck', () => {
      const instance = wrapper.getInstance();

      instance.setState({ words: [{
        english: 'hi',
        spanish: 'hola',
        hint: 'oh-la',
        stack_id: 1,
        isCurrent: false,
        isCompleted: false
      }]});

      expect(wrapper.toJSON()).toMatchSnapshot();
    })

    it('should return current card if there are cards in the deck', () => {
      const instance = wrapper.getInstance();

      instance.setState({ words: [{
        english: 'hi',
        spanish: 'hola',
        hint: 'oh-la',
        stack_id: 1,
        isCurrent: true,
        isCompleted: false
      }]});

      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  })
  
  describe('handleCorrectAnswer', () => {
    
  })
})

