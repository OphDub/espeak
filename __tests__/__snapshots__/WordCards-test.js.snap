// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordCards handleCardLoad should return current card if there are cards in the deck 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B47AEA",
        "display": "flex",
        "height": 63,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "borderColor": "#F5E663",
            "borderRadius": 5,
            "borderWidth": 2,
            "height": 20,
            "marginTop": 15,
            "width": 260,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#F5E663",
              "borderColor": "#F5E663",
              "borderLeftWidth": 2,
              "borderRadius": 2,
              "height": 16,
              "width": "0%",
            }
          }
        />
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "flex-end",
            "color": "#FFF",
            "fontSize": 20,
            "marginLeft": 15,
          }
        }
      >
        50
         pts
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "flex-start",
          "color": "#FFF",
          "marginLeft": 60,
          "marginTop": 5,
        }
      }
    >
      Basics 1
      : 
      0%
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      How do you say...
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "#1E3888",
          "borderRadius": 8,
          "borderWidth": 2,
          "marginVertical": 20,
          "paddingVertical": 35,
          "shadowColor": "#979797",
          "shadowOffset": Object {
            "height": 2,
            "width": 2,
          },
          "shadowOpacity": 0.7,
          "shadowRadius": 4,
          "width": 220,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#1E3888",
            "fontSize": 25,
          }
        }
      >
        HI
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      in Spanish?
    </Text>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Type word in Spanish"
      style={
        Object {
          "borderBottomWidth": 2,
          "borderColor": "#3AAFB9",
          "fontSize": 20,
          "margin": 15,
          "padding": 5,
          "width": 250,
        }
      }
      value=""
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#3AAFB9",
          "borderRadius": 8,
          "marginTop": 10,
          "opacity": 1,
          "padding": 10,
          "shadowColor": "#979797",
          "shadowOffset": Object {
            "height": 2,
            "width": 2,
          },
          "shadowOpacity": 0.7,
          "shadowRadius": 4,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFF",
          }
        }
      >
        SUBMIT
      </Text>
    </View>
  </View>
</View>
`;

exports[`WordCards handleCardLoad should return end screen if there is no current card left in deck 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B47AEA",
        "display": "flex",
        "height": 63,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "borderColor": "#F5E663",
            "borderRadius": 5,
            "borderWidth": 2,
            "height": 20,
            "marginTop": 15,
            "width": 260,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#F5E663",
              "borderColor": "#F5E663",
              "borderLeftWidth": 2,
              "borderRadius": 2,
              "height": 16,
              "width": "0%",
            }
          }
        />
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "flex-end",
            "color": "#FFF",
            "fontSize": 20,
            "marginLeft": 15,
          }
        }
      >
        50
         pts
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "flex-start",
          "color": "#FFF",
          "marginLeft": 60,
          "marginTop": 5,
        }
      }
    >
      Basics 1
      : 
      0%
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#272727",
          "fontSize": 20,
          "marginBottom": 20,
        }
      }
    >
      Great job!
    </Text>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#3AAFB9",
          "borderRadius": 8,
          "opacity": 1,
          "padding": 10,
          "shadowColor": "#979797",
          "shadowOffset": Object {
            "height": 2,
            "width": 2,
          },
          "shadowOpacity": 0.7,
          "shadowRadius": 4,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFF",
          }
        }
      >
        BACK TO MY DECKS
      </Text>
    </View>
  </View>
</View>
`;

exports[`WordCards handleCardLoad should return text saying "No Cards to Render" if no words are stored in state 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B47AEA",
        "display": "flex",
        "height": 63,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "borderColor": "#F5E663",
            "borderRadius": 5,
            "borderWidth": 2,
            "height": 20,
            "marginTop": 15,
            "width": 260,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#F5E663",
              "borderColor": "#F5E663",
              "borderLeftWidth": 2,
              "borderRadius": 2,
              "height": 16,
              "width": "NaN%",
            }
          }
        />
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "flex-end",
            "color": "#FFF",
            "fontSize": 20,
            "marginLeft": 15,
          }
        }
      >
        50
         pts
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "flex-start",
          "color": "#FFF",
          "marginLeft": 60,
          "marginTop": 5,
        }
      }
    >
      Basics 1
      : 
      NaN%
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    No Cards to Render
  </Text>
</View>
`;

exports[`WordCards renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#B47AEA",
        "display": "flex",
        "height": 63,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          Object {
            "borderColor": "#F5E663",
            "borderRadius": 5,
            "borderWidth": 2,
            "height": 20,
            "marginTop": 15,
            "width": 260,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#F5E663",
              "borderColor": "#F5E663",
              "borderLeftWidth": 2,
              "borderRadius": 2,
              "height": 16,
              "width": "NaN%",
            }
          }
        />
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "flex-end",
            "color": "#FFF",
            "fontSize": 20,
            "marginLeft": 15,
          }
        }
      >
        50
         pts
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "flex-start",
          "color": "#FFF",
          "marginLeft": 60,
          "marginTop": 5,
        }
      }
    >
      Basics 1
      : 
      NaN%
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    No Cards to Render
  </Text>
</View>
`;
